Create quiz progress on quiz answers submit

merkushin opened the pull request:
Resolves #7050

We create the quiz progress when start the lesson. When the user submits their answers we implicitly start the lesson.
Here we make sure we have a quiz progress or try to create one.
We don't check the feature flag as it is safe to create quiz progress for comments-based version: it is the same lesson progress there.

## Proposed Changes
* Create quiz progress on lesson progress creation if there is a quiz related to the lesson.

## Testing Instructions
1. Enable `tables_based_progress`.
2. Start a lesson (or go directly to a quiz and submit your answers).
3. Check sensei_lms_progress has quiz progress for the quiz. In case it was created on the lesson start `in-progress` status expected, on the quiz answers submit depends on quiz settings: could be ungraded, passed, failed, graded.


## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [ ] Acceptance criteria is met
- [ ] Decisions are publicly documented
- [ ] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [ ] All strings are translatable (without concatenation, handles plurals)
- [ ] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [ ] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

merkushin committed:
Create quiz progress if there is a quiz in a lesson

merkushin committed:
Test quiz progress created

merkushin committed:
Merge branch 'trunk' into fix/create-quiz-progress-on-submit

merkushin committed:
Fix linter issues

merkushin committed:
Fix undefined Sensei() function

merkushin committed:
Update psalm baseline: remove undefined Sensei() function

merkushin commented on psalm.xml:
```diff
@@ -13,6 +13,9 @@
 		</ignoreFiles>
 	</projectFiles>
+	<stubs>
+		<file name="sensei-lms.php" />
```
We had lots of errors like `Undefined function Sensei()`. I encountered it in this PR as well.
It looks like Psalm for some reason can't discover some entities via reflection. In these cases, I found a recommendation, put files containing those entities into `<stubs>`.
Another approach might be to create `config/psalm/psalm-stubs.php` and define functions like `Sensei()` there (without implementation) and include these stubs in config like I did with above.

m1r0 commented:
Hm, I wasn't able to reproduce this.

I've tried to create a fresh course with a lesson and a quiz. After taking the course and the quiz, it didn't create a quiz progress entry. Did I do something wrong?


https://github.com/Automattic/sensei/assets/1612178/b1c2bf96-e967-405f-83d5-be024b3f31d6

merkushin committed:
Create tables-based quiz progress

merkushin commented:
@m1r0 you're right, my bad. Found a few issues. Looks like I tested on stale data.
Fixed it here: https://github.com/Automattic/sensei/pull/7136/commits/ee6a6289364afdd74b65acc7b02044c10ce46f9c

m1r0 commented:
Thanks, I see the quiz progress but it is added even if the lesson doesn't have a quiz. Is that expected?


https://github.com/Automattic/sensei/assets/1612178/74abaf84-86ee-42c6-9159-d4c3e3a2919f

merkushin committed:
Consider only published quizzes

merkushin commented:
Thanks @m1r0. 
I updated the PR to take into account only published quizzes.
However, I'm still puzzled with a question when the publishing happens.

m1r0 commented:
Nice! Now it creates the quiz progress only when there is a quiz for that lesson. Related to that, does it make more sense to create the quiz progress when you open the quiz itself, similar to how the progress works for lessons?

On a side note, psalm and the linter are failing.

merkushin commented:
> Related to that, does it make more sense to create the quiz progress when you open the quiz itself, similar to how the progress works for lessons?

Makes sense for me. Updated here: https://github.com/Automattic/sensei/pull/7136/commits/fc2bad4c6a707382828d02cc5728524bd110c34f

I removed creation of the quiz progress from the lesson starting function as it leads to a conflict.

merkushin committed:
Create quiz progress on the first visit of the quiz

merkushin committed:
Cast types to int

merkushin committed:
Remove stale test

merkushin committed:
Fix linter issues

merkushin committed:
Add tests for maybe_create_quiz_progress

m1r0 reviewed (approved):
Nice! 💯
